src/irmin/tree.ml

```diff
@@ -1864,91 +1870,59 @@ module Make (P : Private.S) = struct
-        and instead be recomputed one value at a time, using the tag [Blinded]
-        for the values non-loaded in [n], and some other tag for the values
-        loaded in [n]. *)
+        loaded in [n] have been expanded. *)
     and proof_of_node_proof :
         type a. node -> node_proof -> (proof_tree -> a) -> a =
      fun node p k ->
       match p with
       | `Blinded h -> k (Blinded_node h)
-      | `Inode (length, proofs) -> proof_of_inode node length proofs k
-      | `Values vs -> proof_of_values node vs k
+      | `Inode (length, proofs) ->
+          iproof_of_inode node length proofs (fun p -> proof_of_iproof p |> k)
+      | `Values vs -> iproof_of_values node vs (fun p -> proof_of_iproof p |> k)
 
-    and proof_inode_of_node_proof :
+    and iproof_of_node_proof :
         type a. node -> node_proof -> (proof_inode -> a) -> a =
      fun node p k ->
       match p with
       | `Blinded h -> k (Blinded_inode h)
-      | `Inode (length, proofs) -> proof_inode_of_inode node length proofs k
-      | `Values vs -> proof_inode_of_values node vs k
+      | `Inode (length, proofs) -> iproof_of_inode node length proofs k
+      | `Values vs -> iproof_of_values node vs k
 
-    and proof_of_inode :
-        type a. node -> int -> (_ * node_proof) list -> (proof_tree -> a) -> a =
-     fun node length proofs k ->
-      let rec aux acc = function
-        | [] -> k (Inode { length; proofs = List.rev acc })
-        | (index, proof) :: rest ->
-            proof_inode_of_node_proof node proof (fun proof ->
-                aux ((index, proof) :: acc) rest)
-      in
-      match proofs with
-      | [ (index, proof) ] ->
-          proof_inode_of_node_proof node proof (function
-            | Inode_tree { length; proofs = [ (i, p) ] } ->
-                k (Extender { length; segments = [ index; i ]; proof = p })
-            | Inode_extender { length; segments; proof } ->
-                k (Extender { length; segments = index :: segments; proof })
-            | p -> k (Inode { length; proofs = [ (index, p) ] }))
-      | _ -> aux [] proofs
-
-    and proof_inode_of_inode :
+    and iproof_of_inode :
         type a. node -> int -> (_ * node_proof) list -> (proof_inode -> a) -> a
         =
      fun node length proofs k ->
       let rec aux acc = function
         | [] -> k (Inode_tree { length; proofs = List.rev acc })
         | (index, proof) :: rest ->
-            proof_inode_of_node_proof node proof (fun proof ->
+            iproof_of_node_proof node proof (fun proof ->
                 aux ((index, proof) :: acc) rest)
       in
+      (* We are dealing with an inode A.
+         Its children are Bs.
+         The children of Bs are Cs.
+      *)
       match proofs with
       | [ (index, proof) ] ->
-          proof_inode_of_node_proof node proof (function
-            | Inode_tree { length; proofs = [ (i, p) ] } ->
+          (* A is of length 1. *)
```

Comment:
`length` here is the length of `proofs`, not of the inode, I think the comment is a bit confusing